File: examples/canonical-auth/typescript/src/routes/me.ts
import { Hono } from "hono";
import { jwtMiddleware } from "../auth/middleware";
import { AuthContext } from "../auth/context";
import { store } from "../domain/store";
import { getAllowedRoutes } from "../registry/routes";

const app = new Hono();

app.get("/me", jwtMiddleware, async (c) => {
  const auth = c.get("auth") as AuthContext;
  const user = store.getUserById(auth.sub);

  if (!user) {
    return c.json({ error: "User not found" }, 404);
  }

  const tenantName = auth.tenantId ? store.getTenantName(auth.tenantId) : null;
  const isSuperAdmin = auth.tenantRole === "super_admin";
  const allowedRoutes = getAllowedRoutes(auth.permissions, isSuperAdmin);

  const guideSectionsSet = new Set<string>();
  for (const r of allowedRoutes) {
    if (r.description) {
      guideSectionsSet.add(r.description);
    }
  }

  return c.json({
    sub: auth.sub,
    email: auth.email,
    display_name: user.display_name,
    tenant_id: auth.tenantId,
    tenant_name: tenantName,
    tenant_role: auth.tenantRole,
    groups: auth.groups,
    permissions: auth.permissions,
    acl_ver: auth.aclVer,
    allowed_routes: allowedRoutes,
    guide_sections: Array.from(guideSectionsSet),
    issued_at: auth.issuedAt.toISOString().replace(".000Z", "Z"),
    expires_at: auth.expiresAt.toISOString().replace(".000Z", "Z"),
  });
});

export default app;

File: examples/canonical-auth/python/src/routes/me.py
from fastapi import APIRouter, Depends

from src.auth.claims import AuthContext
from src.auth.middleware import get_auth_context, store
from src.registry.routes import get_allowed_routes

router = APIRouter()

@router.get("/me")
def get_me(auth: AuthContext = Depends(get_auth_context)):
    user = store.get_user_by_id(auth.sub)
    if not user:
        # Should be caught by middleware but safe to check
        return {"error": "User not found"}
        
    tenant_name = store.get_tenant_name(auth.tenant_id) if auth.tenant_id else None
    
    is_super_admin = auth.tenant_role == "super_admin"
    allowed_routes = get_allowed_routes(auth.permissions, is_super_admin)
    
    guide_sections = list(set([r.get("description", "") for r in allowed_routes if "description" in r]))
    
    return {
        "sub": auth.sub,
        "email": auth.email,
        "display_name": user.display_name,
        "tenant_id": auth.tenant_id,
        "tenant_name": tenant_name,
        "tenant_role": auth.tenant_role,
        "groups": auth.groups,
        "permissions": auth.permissions,
        "acl_ver": auth.acl_ver,
        "allowed_routes": allowed_routes,
        "guide_sections": guide_sections,
        "issued_at": auth.issued_at.isoformat().replace("+00:00", "Z"),
        "expires_at": auth.expires_at.isoformat().replace("+00:00", "Z")
    }
